File: DuDoong-Domain/src/main/java/band/gosrock/domain/domains/order/service/handler/WithDrawOrderHandler.java
package band.gosrock.domain.domains.order.service.handler;


import band.gosrock.domain.common.events.order.WithDrawOrderEvent;
import band.gosrock.domain.domains.order.adaptor.OrderAdaptor;
import band.gosrock.domain.domains.order.domain.Order;
import band.gosrock.domain.domains.order.domain.OrderStatus;
import band.gosrock.domain.domains.order.service.WithdrawPaymentService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

@Component
@RequiredArgsConstructor
@Slf4j
public class WithDrawOrderHandler {

    private final WithdrawPaymentService withdrawPaymentService;

    private final OrderAdaptor orderAdaptor;

    @Async
    @TransactionalEventListener(
            classes = WithDrawOrderEvent.class,
            phase = TransactionPhase.AFTER_COMMIT)
    @Transactional
    public void handleRegisterUserEvent(WithDrawOrderEvent withDrawOrderEvent) {
        log.info(withDrawOrderEvent.getOrderUuid() + "주문 철회 핸들러");
        OrderStatus orderStatus = withDrawOrderEvent.getOrderStatus();
        Order order = orderAdaptor.findByOrderUuid(withDrawOrderEvent.getOrderUuid());
        String reason = "결제 취소";
        if (orderStatus.equals(OrderStatus.CANCELED)) {
            reason = "이벤트 관리자에 의한 취소";

        } else if (orderStatus.equals(OrderStatus.REFUND)) {
            reason = "구매자에의한 환불 요청";
        }

        log.info(withDrawOrderEvent.getOrderUuid() + "주문 철회" + order.getPaymentKey() + reason);
        withdrawPaymentService.execute(order.getUuid(), order.getPaymentKey(), reason);
    }
}

File: DuDoong-Domain/src/main/kotlin/band/gosrock/domain/domains/order/service/handler/WithDrawOrderHandler.kt
package band.gosrock.domain.domains.order.service.handler

import band.gosrock.domain.common.events.order.WithDrawOrderEvent
import band.gosrock.domain.domains.order.adaptor.OrderAdaptor
import band.gosrock.domain.domains.order.domain.OrderStatus
import band.gosrock.domain.domains.order.service.WithdrawPaymentService
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

@Component
class WithDrawOrderHandler(
    private val withdrawPaymentService: WithdrawPaymentService,
    private val orderAdaptor: OrderAdaptor,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Async
    @TransactionalEventListener(classes = [WithDrawOrderEvent::class], phase = TransactionPhase.AFTER_COMMIT)
    fun handleWithDrawOrderEvent(withDrawOrderEvent: WithDrawOrderEvent) {
        log.info("${withDrawOrderEvent.orderUuid} 주문 철회 핸들러")
        val orderStatus = withDrawOrderEvent.orderStatus
        val order = orderAdaptor.findByOrderUuid(withDrawOrderEvent.orderUuid)
        if (!order.isPaid()) return

        val reason = when (orderStatus) {
            OrderStatus.CANCELED -> "이벤트 관리자에 의한 취소"
            OrderStatus.REFUND -> "구매자에의한 환불 요청"
            else -> "결제 취소"
        }

        log.info("${withDrawOrderEvent.orderUuid} 주문 철회 ${order.paymentKey}$reason")
        withdrawPaymentService.execute(withDrawOrderEvent.orderUuid, order.paymentKey, reason)
    }
}
